### PR TITLE
61 handle bean creation in bringcontainer class by single constructor with arguments

### LINKED ISSUE #61: Handle bean creation in BringContainer class by single constructor with arguments
`BeanDefinitionReaderUtils` should look for single constructor. Constructor could have any parameters.
`SimpleObjectFactory` should take single constructor dependencies from bean definition and create bean out of this information. If there is some dependent bean missing, then this bean should be created first, and so on.

### PR DESCRIPTION
**Pull request by JoshukLite**
Created logic to handle the creation of beans with a single constructor having 0 or more parameters.
Adapted BeanDefinition logic to handle class type along with bean name.
Added handling of some of the edge cases - bean name duplicate case, number of bean constructors case, failed bean creation case.
Covered BeanDefinitionReaderUtils and BringContainer with tests.

### CODE REVIEW

**Review by golichenko02 — APPROVED** [review]

**Review by s-barabash — APPROVED** [review]

File: src/main/java/io/github/bobocodebreskul/context/registry/AnnotatedBeanDefinitionReader.java [hunk 1]
```diff
@@ -121,10 +122,10 @@ private <T> void doRegisterBean(Class<T> beanClass, String name) {
       annotatedBeanDefinition.setPrimary(true);
     }
 
-    List<Class<?>> beanDependencies = getBeanDependencies(beanClass);
+    List<BeanDependency> dependencies = getBeanDependencies(beanClass);
```
Comment (mykolaFilimonov) [inline src/main/java/io/github/bobocodebreskul/context/registry/AnnotatedBeanDefinitionReader.java:125]: maybe better to do call through the Util class not do import on this

Comment (mykolaFilimonov) [reply src/main/java/io/github/bobocodebreskul/context/registry/AnnotatedBeanDefinitionReader.java:125]: it will look more clear what happened and how

File: src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java [hunk 2]
```diff
@@ -35,32 +38,43 @@ public static String generateBeanName(BeanDefinition beanDefinition,
       log.error("Failed to generate bean name for nullable bean definition");
       return "Null bean definition specified";
     });
-    Objects.requireNonNull(beanDefinition.getBeanClass(), () -> {
-      log.error("Failed to generate bean name for nullable bean class");
-      return "Bean class has not been specified";
-    });
+    validateBeanClassNonNull(beanDefinition.getBeanClass());
+    validateBeanClassName(beanDefinition, registry);
```
Comment (mykolaFilimonov) [inline src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java:42]: I think we do not need to send the registry here, and maybe better to do validation on the reader level, because it couples us with implementation which says we required to have some registry

File: src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java [hunk 3]
```diff
@@ -35,32 +38,43 @@ public static String generateBeanName(BeanDefinition beanDefinition,
...
+  public static List<BeanDependency> getBeanDependencies(Class<?> beanClass) {
     log.trace("Scanning class {} for @Autowire candidates", beanClass.getName());
-    Stream<Class<?>> constructorDependenciesStream = Arrays.stream(
-            beanClass.getDeclaredConstructors())
+
+    Constructor<?>[] declaredConstructors = beanClass.getDeclaredConstructors();
+
+    validateBeanClassConstructorSize(beanClass.getName(), declaredConstructors);
+
+    Stream<Class<?>> constructorDependenciesStream = Arrays.stream(declaredConstructors)
```
Comment (mykolaFilimonov) [inline src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java:77]: Do we need stream there? if we expect exactly one constructor currently maybe it redundant there and add over logic and miss understanding for reader

File: src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java [hunk 4]
```diff
@@ -73,7 +87,8 @@ public static List<Class<?>> getBeanDependencies(Class<?> beanClass) {
 
     return Stream.concat(Stream.concat(constructorDependenciesStream, fieldDependenciesStream),
             methodDependenciesStream)
-        .collect(Collectors.toList());
+        .map(dependency -> new BeanDependency(generateClassBeanName(dependency), dependency))
```
Comment (mykolaFilimonov) [inline src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java:90]: may it broke something if we remove other dependencies except constractor from concat method? Because in case if add @Autowired somewhere it looks like it added additional dependencies which will broke logic from Container

File: src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java [hunk 5]
```diff
@@ -90,19 +105,48 @@ public static boolean isBeanAutowireCandidate(Class<?> beanClass,
     log.trace("Check if bean class {} is autowire candidate", beanClass.getName());
     boolean isAutowiredCandidate = registry.getBeanDefinitions().stream()
         .filter(beanDefinition -> !beanDefinition.getBeanClass().equals(beanClass))
-        .flatMap(beanDefinition -> beanDefinition.getDependsOn().stream())
-        .anyMatch(beanDefinitionClass -> beanDefinitionClass.equals(beanClass));
+        .flatMap(beanDefinition -> beanDefinition.getDependencies().stream())
+        .anyMatch(beanDefinitionClass -> beanDefinitionClass.type().equals(beanClass));
```
Comment (mykolaFilimonov) [inline src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java:109]: if we decide to use registry we need to document this

**Review by mykolaFilimonov — CHANGES_REQUESTED** [review]
just small question about code, if it not for this ticket we may create small enhancment issues for future

File: src/main/java/io/github/bobocodebreskul/context/registry/AnnotatedBeanDefinitionReader.java (hunk 1 above)
Comment (golichenko02) [reply src/main/java/io/github/bobocodebreskul/context/registry/AnnotatedBeanDefinitionReader.java:125]: We use the static import from this class for other methods either, the purpose of this is a concise and readable code.
There's a debug and trace level logs in these utility method calls, which should give a clear understanding what happened and where.
Did I convince you?)

File: src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java (hunk 3 above)
Comment (golichenko02) [reply src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java:77]: This is old changes that would be changed in the next pull request after this one will have been merged. It will require some additional refactoring changes, so I propose  not to do them here.

File: src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java (hunk 4 above)
Comment (golichenko02) [reply src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java:90]: I'll do it in scope of this task https://github.com/bobocode-breskul/bring/issues/66

File: src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java (hunk 5 above)
Comment (golichenko02) [reply src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java:109]: can you clarify this point please? 
registry is used here to retrieve all registered bean definitions and we documented it in javadoc

File: src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java (hunk 2 above)
Comment (golichenko02) [reply src/main/java/io/github/bobocodebreskul/context/support/BeanDefinitionReaderUtils.java:42]: I assumed that any implementation will require bean definition registry, isn't?
Otherwise we need to refactor everything :(

File: src/main/java/io/github/bobocodebreskul/context/registry/AnnotatedBeanDefinitionReader.java (hunk 1 above)
Comment (mykolaFilimonov) [reply src/main/java/io/github/bobocodebreskul/context/registry/AnnotatedBeanDefinitionReader.java:125]: not fully but ok

**Review by mykolaFilimonov — APPROVED** [review]
LGTM
